extension/duckdb/src/duckdb_connector.cpp

```diff
@@ -0,0 +1,93 @@
+#include "duckdb_connector.h"
+
+#include "common/exception/runtime.h"
+#include "common/file_system/virtual_file_system.h"
+#include "main/client_context.h"
+
+namespace kuzu {
+namespace duckdb_extension {
+
+static DuckDBConnectionType getConnectionType(const std::string& path) {
+    if (path.rfind("https://", 0) == 0 || path.rfind("http://", 0) == 0) {
+        return DuckDBConnectionType::HTTP;
+    } else if (path.rfind("s3://", 0) == 0) {
+        return DuckDBConnectionType::S3;
+    } else {
+        return DuckDBConnectionType::LOCAL;
+    }
+}
+
+std::unique_ptr<duckdb::MaterializedQueryResult> DuckDBConnector::executeQuery(
+    std::string query) const {
+    KU_ASSERT(instance != nullptr && connection != nullptr);
+    auto result = connection->Query(query);
+    if (result->HasError()) {
+        throw common::Exception{result->GetError()};
+    }
+    return result;
+}
+
+void LocalDuckDBConnector::connect(const std::string& dbPath, const std::string& /*catalogName*/,
+    main::ClientContext* context) {
+    if (!context->getVFSUnsafe()->fileOrPathExists(dbPath, context)) {
+        throw common::RuntimeException{
+            common::stringFormat("'{}' is not a valid duckdb database path.", dbPath)};
+    }
+    duckdb::DBConfig dbConfig{true /* isReadOnly */};
+    instance = std::make_unique<duckdb::DuckDB>(dbPath, &dbConfig);
+    connection = std::make_unique<duckdb::Connection>(*instance);
+}
+
+void HTTPDuckDBConnector::connect(const std::string& dbPath, const std::string& catalogName,
+    main::ClientContext* /*context*/) {
+    // Creates an in-memory duckdb instance, then install httpfs and attach remote duckdb.
+    instance = std::make_unique<duckdb::DuckDB>(nullptr);
+    connection = std::make_unique<duckdb::Connection>(*instance);
+    executeQuery("install httpfs;");
+    executeQuery("load httpfs;");
+    executeQuery(common::stringFormat("attach '{}' as {} (read_only);", dbPath, catalogName));
```

Comment:
I'm a bit lost on the variable `catalogName` here, isn't this database alias?

Reply:
Yes, duckdb uses the term catalogName